python/lsst/ip/diffim/imageDecorrelation.py

```diff
@@ -196,235 +196,239 @@ def run(self, exposure, templateExposure, subtractedExposure, psfMatchingKernel,
+        """Compute the Lupton decorrelation post-conv. kernel for decorrelating an
+        image difference, based on the PSF-matching kernel.
 
         Parameters
         ----------
-        kernel : `numpy.array`
-            a numpy.array
+        kappa : `numpy.ndarray`
+            A matching kernel 2-d numpy.array derived from Alard & Lupton PSF matching
+        svar : `float`, optional
+            Average variance of science image used for PSF matching
+        tvar : `float`, optional
+            Average variance of the template (matched) image used for PSF matching
+        preConvArr : `numpy.ndarray`, optional
+            If not None, then pre-filtering was applied
+            to science exposure, and this is the pre-convolution kernel.
 
         Returns
         -------
-        out : `numpy.array`
-            a fixed kernel numpy.array. Returns a copy if the dimensions needed to change;
-            otherwise just return the input kernel.
+        corrft : `numpy.ndarray` of complexes
+            The frequency space representation of the correction. The array is of complex
+            dtype but contains all real numbers. Shape is `self.freqSpaceShape`.
+
+        Notes
+        -----
+        The maximum correction factor converges to `sqrt(tvar/svar)` towards high frequencies.
+        This should be a plausible value.
         """
-        # Note this works best for the FFT if we left-pad
-        out = kernel
-        changed = False
-        if (out.shape[0] % 2) == 1:
-            out = np.pad(out, ((1, 0), (0, 0)), mode='constant')
-            changed = True
-        if (out.shape[1] % 2) == 1:
-            out = np.pad(out, ((0, 0), (1, 0)), mode='constant')
-            changed = True
-        if changed:
-            out *= (np.mean(kernel) / np.mean(out))  # need to re-scale to same mean for FFT
-        return out
+        kappa = self.padCenterOriginArray(kappa, self.freqSpaceShape)
+        kft = np.fft.fft2(kappa)
+        kft2 = np.conj(kft) * kft
+        if preConvArr is None:
+            denom = svar + tvar * kft2
+        else:
+            preConvArr = self.padCenterOriginArray(preConvArr, self.freqSpaceShape)
+            mk = np.fft.fft2(preConvArr)
+            mk2 = np.conj(mk) * mk
+            denom = svar * mk2 + tvar * kft2
+        kft = np.sqrt((svar + tvar) / denom)
+        return kft
 
-    @staticmethod
-    def _fixEvenKernel(kernel):
-        """Take a kernel with even dimensions and make them odd, centered correctly.
+    def computeCorrectedDiffimPsf(self, corrft, psfArr):
+        """Compute the (decorrelated) difference image's new PSF.
 
         Parameters
         ----------
-        kernel : `numpy.array`
-            a numpy.array
+        corrft : `numpy.ndarray`
+            The frequency space representation of the correction calculated by
+            `computeCorrection`. Shape must be `self.freqSpaceShape`.
+        psfArr : `numpy.ndarray`
```

Comment:
`psfArr` made me think of an array of psfs at first. Perhaps `psfVals` or `psfOld`? Also, the current implementation modifies `psfArr` in place (as well as returning it), so you should document that if you leave that behavior.

Reply:
`padCenterOriginArray` always creates a copy.

Reply:
However `computeCorrectedImage` indeed made an unintended in place modification...